src/common/src/types/mod.rs

```diff
@@ -629,6 +629,23 @@ impl ToDatumRef for DatumRef<'_> {
     }
 }
 
+/// To make sure there is `as_scalar_ref` for all scalar ref types.
+pub trait SelfAsScalarRef {
+    fn as_scalar_ref(&self) -> Self;
+}
```

Comment:
Why we need this?

Reply:
The purpose is to allow both owned type and reference type as function return value. (e.g. `-> &str` and `-> Box<str>`). So we want both types have a method `.as_scalar_ref() -> &str` so that we can call it to convert owned type into reference and keep reference type unchanged in the macro. Currently this method exists in the `Scalar` trait. At first I added this method to the `ScalarRef` trait, but it leads to conflicts for those types that share both traits (e.g. primitive types: i32, i64...). Therefore I created this trait to add `as_scalar_ref` for types which `impl ScalarRef` but not `impl Scalar`. Maybe a better solution is to extract an `AsScalarRef` trait. But it's too hard for me to do type exercise. 🤪

Reply:
Cool